### PR TITLE
Add possibility to cut on timing to IVF (93X)

### PR DESCRIPTION
**Pull request by lgray**
Backport of #20640.

### CODE REVIEW

**Comment by lgray** [comment]
@cmsbuild please test

**Comment by kpedro88** [comment]
assign upgrade

**Comment by lgray** [comment]
@cmsbuild please test

**Comment by perrotta** [comment]
please test

**Comment by lgray** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/21193	990d867
- cuts on track and vertex time in IVF, enabled only in workflows with phase-2 timing. The backport has identical changes to what's in the master branch PR https://github.com/cms-sw/cmssw/pull/20640	7a501f1
- jenkins tests pass and comparisons show  changes only in 20434 workflow (ignoring existing random diffs)
- see more notes on performance changes in the signoff notes in the master branch https://github.com/cms-sw/cmssw/pull/20640#issuecomment-344551808

**Comment by kpedro88** [comment]
+1

**Comment by slava77** [comment]
I've masked out my +1
I've reminded myself of #20621 : this PR will make b-tagging results non-reproducible, at least in multithreaded jobs. This seems like a showstopper for nominal production.

**Comment by lgray** [comment]
Hi Slava that only happens when the vertices have timing, otherwise it's
not used.

**Comment by slava77** [comment]
On 11/16/17 4:10 PM, Lindsey Gray wrote:
> Hi Slava that only happens when the vertices have timing, otherwise it's
> not used.

This [unreproducible b-tagging] is the behavior of this PR out of the
box for D19/timing workflows.

>

**Comment by slava77** [comment]
I checked with 20034.2 (D17 with timing) and the comparison of two runs (one single thread and one multithread to make event reordering happen) show that B-tagging plots change.
The "D17Timing" is the currently standard production workflow
e.g. https://cmsweb.cern.ch/couchdb/reqmgr_config_cache/23561dc2f1da997a35cfb1e3ca22a16b/configFile for /TT_TuneCUETP8M2T4_14TeV-powheg-pythia8/PhaseIITDRFall17DR-PU200_93X_upgrade2023_realistic_v2-v1

**Comment by lgray** [comment]
ok - that's not good to begin with, D17 shouldn't change. I'll figure it
out.

**Comment by slava77** [comment]
IVF clustering and sorting uses track times when available, regardless of
PV. That's the origin of changes.

**Comment by slava77** [comment]
So, to iterate on the reason of this PR to be held:
the expectation is that the D17+timing workflow, which is the production workflow for HGCAL,
is that there should be no change in b-tagging variables at all.
Previously mentioned non-reproducibility is mostly secondary.

**Comment by lgray** [comment]
@slava77 Yes - I am working on it today!

**Comment by slava77** [comment]
Based on the discussion in the ORP on Nov 21, this PR is required only for the MTD production, which will apparently happen in a release to be made some time after the HGCAL miniAOD release is made.

**Comment by slava77** [comment]
@lgray please clarify on the status of this PR

**Comment by slava77** [comment]
@lgray please clarify on the status of this PR
Are you going to make updates or should we close this?

**Comment by lgray** [comment]
Hi Slava - I'll work on this today. Do not close.

**Comment by lgray** [comment]
@cmsbuild please test

**Comment by lgray** [comment]
@slava77 This should remove the issue.

**Comment by slava77** [comment]
>@slava77 This should remove the issue.

Please submit these changes to the master branch. That's where they will be integrated first.

**Comment by slava77** [comment]
B-tag plots are still changing in 20434.0 (it is a D19 workflow, but in it the default PV is still the 3D and IIUC, it should behave the same as the 20034.2 D17Timing).
Please change the PR to apply the cut based on configuration instead of per-event content heuristics.

File: RecoVertex/AdaptiveVertexFinder/interface/SVTimeHelpers.h [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    const auto& trks = vtx.originalTracks();
+    double meantime = 0., expv_x2 = 0., normw = 0., timecov = 0.;		  
+    for( const auto& trk : trks ) {
+      if( edm::isFinite(trk.timeExt()) ) {
+	const double time = trk.timeExt();
+	const double inverr = 1.0/trk.dtErrorExt();
+	const double w = inverr*inverr;
+	meantime += time*w;
+	expv_x2  += time*time*w;
+	normw    += w;
```
Comment (slava77) [inline RecoVertex/AdaptiveVertexFinder/interface/SVTimeHelpers.h]: I think that this is another place where time of tracks contributes to the SV fit bypassing the PV.cov33>0 check

**Comment by slava77** [comment]
@lgray 
please check the latest comments here.
It looks like this PR is still incomplete.

**Comment by lgray** [comment]
Yes, I will try to get to them this week.

**Comment by slava77** [comment]
@davidlange6 
please reopen this.
This PR is the only simple way to test the fix.

Testing the fix in master requires local tests with a non-standard baseline with a revert of the feature already merged in 10X .. which couldn't be reverted easily in #21691

**Comment by slava77** [comment]
@lgray 
please clarify if there was some progress with this PR
Thank you.

**Comment by slava77** [comment]
[a weekly ping]
@lgray
please clarify if there was some progress with this PR
Thank you.

**Comment by slava77** [comment]
[a weekly ping]
@lgray
please clarify if there was some progress with this PR
Thank you.

**Comment by slava77** [comment]
@kpedro88 
should we move the fix of this problem to an issue that may or may not be fixed any time soon?

**Comment by kpedro88** [comment]
I don't think we'll continue to use 93X for Phase2 anymore, so this PR at least can be closed. Opening an issue is probably a good idea.

**Comment by slava77** [comment]
On 1/18/18 8:23 AM, Kevin Pedro wrote:
> I don't think we'll continue to use 93X for Phase2 anymore, so this PR
> at least can be closed. Opening an issue is probably a good idea.

the question applies equivalently to #21659
It's not about 93X. I asked this PR to be kept open because it's the
easiest to test the problem here with jenkins.
In the master the test will require making a baseline that reverts the
#20640
to see that the fix is correct.

So, is the answer the same for #21659 that it can be closed?
If no, then I'd like to keep this open to simplify testing.

**Comment by kpedro88** [comment]
In that case let's leave it open. Just pinged @lgray and he hopes to have some time soon.

**Comment by slava77** [comment]
[a weekly ping]
@lgray
please clarify if there was some progress with this PR
Thank you.

**Comment by slava77** [comment]
[a weekly ping]
@lgray
please clarify if there was some progress with this PR
Thank you.

**Comment by slava77** [comment]
[a weekly ping]
@lgray
please clarify if there was some progress with this PR
Thank you.

**Comment by slava77** [comment]
-1

the issue is to be followed up in #22185

**Comment by lgray** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for  https://github.com/cms-sw/cmssw/pull/21193	89c23cc
- changes are in line with the PR description; this PR is in agreement with what was integrated in the master branch with PR #20640 on Nov 16 together with the long-awaited bugfix in #21659. The overall backport is OK.
- jenkins tests pass and comparisons with the baseline show no differences
    - one somewhat significant caveat is that due to the significant change in the data format the default setup of the fwlite-based reco tests does not work and only the DQM plots can be used. The DQM comparisons should be enough in this context.

**Pull request closed without merging**